api/monitor/monitorclient/client.go

```diff
@@ -0,0 +1,217 @@
+type MonitorOption func(m *MonitorClient)
+
+// WithConn adds the given conns to the MonitorClient
+func WithConn(name string, conn *grpc.ClientConn) MonitorOption {
+	return func(m *MonitorClient) {
+		m.clients[name] = monitor.NewMonitorClient(conn)
+	}
+}
+
+// WithServer [DEPRECATED] connects to the given server and adds it to the MonitorClient
+// Instead of using WithServer, you should set up the gRPC connection externally and use
+// WithConn to add it to the MonitorClient.
+func WithServer(name, addr string, opts ...grpc.DialOption) MonitorOption {
+	if len(opts) == 0 {
+		if strings.HasSuffix(name, "-tls") {
+			netHost, _, _ := net.SplitHostPort(addr)
+			creds := credentials.NewTLS(pool.TLSConfig(netHost))
+			opts = append(opts, grpc.WithTransportCredentials(creds))
+		} else {
+			opts = append(opts, grpc.WithInsecure())
+		}
+	}
+	return func(m *MonitorClient) {
+		conn, err := grpc.Dial(addr, opts...)
+		if err != nil {
+			return
+		}
+		m.clients[name] = monitor.NewMonitorClient(conn)
+	}
+}
+
+// WithStatusTicker returns a MonitorOption that sets a status ticker
+func WithStatusTicker(interval time.Duration) MonitorOption {
+	return func(m *MonitorClient) {
+		m.statusTicker = time.NewTicker(interval)
+	}
+}
+
+// NewMonitorClient returns a new MonitorClient
+func NewMonitorClient(opts ...MonitorOption) *MonitorClient {
+	m := &MonitorClient{
+		log:        log.Get(),
+		bufferSize: DefaultBufferSize,
+		clients:    make(map[string]monitor.MonitorClient),
+	}
+	for _, opt := range opts {
+		opt(m)
+	}
+	return m
+}
+
+// MonitorClient is an overlay on top of the gRPC client
+type MonitorClient struct {
+	log          log.Interface
+	bufferSize   int
+	clients      map[string]monitor.MonitorClient
+	statusTicker *time.Ticker
+}
+
+// TickStatus calls f periodically if the StatusTicker has been set
+func (m *MonitorClient) TickStatus(f func()) {
+	if m.statusTicker == nil {
+		return
+	}
+	for range m.statusTicker.C {
+		f()
+	}
+}
+
+// Stream interface allows sending anything with Send()
+type Stream interface {
+	Send(msg interface{})
+	Open()
+	Close()
+	Reset()
+}
+
+type componentClient struct {
+	log log.Interface
+	mu  sync.RWMutex
+
+	status   []*sendbuffer.Stream
+	uplink   []*sendbuffer.Stream
+	downlink []*sendbuffer.Stream
+
+	setup  func()
+	cancel context.CancelFunc
+}
+
+func (c *componentClient) Open() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.cancel != nil {
+		return
+	}
+	c.setup()
+}
+
+func (c *componentClient) Close() {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	if c.cancel == nil {
+		return
+	}
+	c.cancel()
+}
+
+func (c *componentClient) Reset() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.cancel == nil {
+		return
+	}
+	c.cancel()
+	c.setup()
+}
+
+func (c *componentClient) Send(msg interface{}) {
+	switch msg := msg.(type) {
+	case *gateway.Status, *router.Status, *broker.Status, *networkserver.Status, *handler.Status:
+		if len(c.status) == 0 {
+			return
+		}
+		for _, cli := range c.status {
```

Comment:
`cli` is the convention in `ttn`, and now is not the time to change those conventions